### PR TITLE
[W5.4][T12-A3]Grant Miller

### PR DESCRIPTION
**Pull request by grantcm**
Implemented a Birthday field for the add command. Updated JUnit Tests accordingly.

### CODE REVIEW

File: src/seedu/addressbook/commands/DeleteCommand.java [hunk 1]
```diff
@@ -1,9 +1,12 @@
 package seedu.addressbook.commands;
 
 import seedu.addressbook.common.Messages;
+import seedu.addressbook.data.AddressBook;
 import seedu.addressbook.data.person.ReadOnlyPerson;
 import seedu.addressbook.data.person.UniquePersonList.PersonNotFoundException;
 
+import java.util.List;
```
Comment (eugenepeh) [inline src/seedu/addressbook/commands/DeleteCommand.java]: should arrange in alphabetical order, e.g. `import java.util.List;` should be on top of `import seedu.addressbook.common.Messages;`

File: src/seedu/addressbook/data/AddressBook.java [hunk 2]
```diff
@@ -1,16 +1,14 @@
 package seedu.addressbook.data;
 
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (eugenepeh) [inline src/seedu/addressbook/data/AddressBook.java]: remember to disable intellij auto import wildcard
[Why is wild card import bad?](https://stackoverflow.com/questions/147454/why-is-using-a-wild-card-with-a-java-import-statement-bad)

File: src/seedu/addressbook/data/person/Contact.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package seedu.addressbook.data.person;
+
+public class Contact implements Printable{
```
Comment (eugenepeh) [inline src/seedu/addressbook/data/person/Contact.java]: space before left curly should be consistent

File: src/seedu/addressbook/data/person/Person.java [hunk 4]
```diff
@@ -55,10 +60,30 @@ public Address getAddress() {
     }
 
     @Override
+    public Birthday getBirthday() {return  birthday; }
```
Comment (eugenepeh) [inline src/seedu/addressbook/data/person/Person.java]: keep single statement per line; `return birthday;` deserve a line on its own

File: src/seedu/addressbook/data/person/Person.java [hunk 5]
```diff
@@ -55,10 +60,30 @@ public Address getAddress() {
...
+    }
+
+    /**
+     * Returns a concatenated version of the printable strings of each object.
+     */
+    public String getPrintableString(Printable... printables){
+        String output="";
+
+        for(Printable p:printables){
+            output+=p.getPrintableString()+" ";
```
Comment (eugenepeh) [inline src/seedu/addressbook/data/person/Person.java]: include spaces before and after operators to maintain readability.
also can consider using StringBuilder for better efficiency or lambda for better readability.

File: src/seedu/addressbook/data/person/Person.java [hunk 6]
```diff
@@ -76,12 +101,14 @@ public boolean equals(Object other) {
...
         // use this method for custom fields hashing instead of implementing your own
-        return Objects.hash(name, phone, email, address, tags);
+        return Objects.hash(name, phone, email, address, birthday, tags);
     }
 
     @Override
     public String toString() {
         return getAsTextShowAll();
+        //Using LO Interface Segregation Principle
+        //return getPrintableString(name,phone,email,address,birthday);
```
Comment (eugenepeh) [inline src/seedu/addressbook/data/person/Person.java]: don't keep commented codes, there's stash and revision for that

File: src/seedu/addressbook/data/person/Email.java [hunk 7]
```diff
@@ -38,24 +34,10 @@ public static boolean isValidEmail(String test) {
...
-    @Override
     public boolean equals(Object other) {
         return other == this // short circuit if same object
                 || (other instanceof Email // instanceof handles nulls
                 && this.value.equals(((Email) other).value)); // state check
     }
 
-    @Override
-    public int hashCode() {
-        return value.hashCode();
```
Comment (eugenepeh) [inline src/seedu/addressbook/data/person/Email.java]: [Why you shouldn't remove hashCode()](http://crunchify.com/how-to-override-equals-and-hashcode-method-in-java/)

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 8]
```diff
@@ -10,6 +10,23 @@
 
 public class UtilsTest {
 
+    @Test
+     public void isAnyNull() throws Exception {
+
+
+        // non empty list with nulls in the middle
+        assertTrue(Utils.isAnyNull(new Object(), null, null, "test"));
+        assertTrue(Utils.isAnyNull("", null, new Object()));
```
Comment (eugenepeh) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Is checking the values in the middle necessary?
Not covered yet but for your reference [Boundary value analysis](https://www.google.com.sg/url?sa=t&rct=j&q=&esrc=s&source=web&cd=2&cad=rja&uact=8&ved=0ahUKEwiC5rGmvqTWAhVHNo8KHXUIDBEQFggqMAE&url=https%3A%2F%2Fen.wikipedia.org%2Fwiki%2FBoundary_testing&usg=AFQjCNGq35j0mszLB324NqJzWN2qKJYNLg)
